### ISSUE TITLE
Is there a complete example config for Webpack 2?

### ISSUE THREAD

**Issue by dschissler** [webpack/webpack#2782]
I'm having trouble getting my first loader to work while converting to Webpack 2.  I cannot find a simple webpack 2 config for setting up Babel and few common plugins and loaders.  I've followed the whats new gist post and I've made many changes.  The rules say not to post here about questions and tech support but its July and I can't get a response anywhere else in two days now.

**Comment by dschissler** [comment]
I'm using `2.1.0-beta.19`.

Here is my resolve:

```javascript
  resolve: {
    modules: [appModulesRoot, nodeModulesRoot, bowerRoot, themeRoot],
    descriptionFiles: ['package.json', 'bower.json'],
    mainFields: ['main', 'browser'],
    mainFiles: ['index'],
    extensions: [
      '',
      '.js', '.coffee',
      '.html', '.njk',
      '.css', '.scss', '.less',
      '.json', '.yml'
    ],
    enforceExtension: false,
    moduleExtensions: ['-loader'],
    enforceModuleExtension: false,
    alias: {
      underscore: 'lodash',
      handlebars: 'handlebars/dist/handlebars',
      highlight: 'highlight.js/lib/highlight'
    },
  },
```

I'm still using Babel 5 on this project and I'm planning to upgrade once I get Webpack 2 working to some degree:

```javascript
      {
        test: /\.js$/,
        exclude: /(node_modules|bower_components)/,
        loader: 'babel-loader',
        query: {
          optional: ['runtime'],
          stage: 0,
          cacheDirectory: '/tmp'
        }
      }, {
```

I'm getting a lot of messages like this: `ERROR in Entry module not found: Error: Can't resolve 'babel-loader?{"optional":["runtime"],"stage":0,"cacheDirectory":"/tmp"}' in '/opt/project/app/webpack'`

**Comment by sokra** [comment]
Loaders resolving is not affected by the `resolve` key. Loaders resolving has a separate key: `resolveLoader` (with the same content).

**Comment by dschissler** [comment]
In terms of my config above could you give me a one liner or direct me to the setting that I have wrong?

**Comment by sokra** [comment]
Try `--display-error-details` first...

Does `/opt/project/app/webpack/node_modules/babel-loader` exist?

**Comment by dschissler** [comment]
Sorry the the snap weak reply.  I thought about it and then realized that I need `resolveLoader` as that was commented out.  Then to start with I copied all of the settings from `resolve` into `resolveLoader` and I'll clean it up after its working.

Now it seems I need to get the exports and expose working.  I'm maintaining a test bed for Phalcon and Webpack integration and it uses some old JS tech that is funky.

```
      {
        test: /[\/]angular\.js$/,
        loader: 'exports',
        query: {
          angular: true
        }
      }, {
        test: /jquery\.js$/,
        loader: 'expose',
        query: {
          jQuery: true,
          $: true
        }
      }, {
```

**Comment by dschissler** [comment]
To clarify `resolveLoader` settings weren't there and now its finding loaders but I have this issue with these simple loaders now.

**Comment by dschissler** [comment]
@sokra since it appears that the `query: {}` mode of passing options works for Webpack 1 & 2 and that it is discontinued for 2 that the simple examples in the loaders should be made to use only the new way.

**Comment by d3viant0ne** [comment]
@dschissler - My suggestion would be to start here https://github.com/gdi2290/angular2-webpack2-starter

@gdi2290's Webpack2 example has a pretty complete dev config.

**Comment by dschissler** [comment]
@d3viant0ne Thanks, it might be useful later but I don't see there how to setup the `expose-loader` and `exports-loader` for Webpack 2.  I'm just having trouble converting it from the query string format to the query object.

**Comment by d3viant0ne** [comment]
Yeah, sorry. Read the first post and then "skimmed" the rest :/

**Comment by dschissler** [comment]
I still need an example for exports and expose loader setup with Webpack 2.  It should be real simple.  I just need the new `query` syntax for it.

**Comment by d3viant0ne** [comment]
@dschissler - you have this project public somewhere. I'd like to tinker with it and see if I can work out your query object problem.

**Comment by dschissler** [comment]
Thanks.  I haven't committed this work since it is broken but you can see the Webpack 1 version here: https://github.com/perchlayer/webird/blob/master/dev/gulpfile.webpack.js#L122-L126

**Comment by nareshbhatia** [comment]
I am also switching over to webpack 2. I have a complete starter project with React and Redux over here: https://github.com/sapientglobalmarkets/react-redux-seed. However, like everyone else, I am struggling through the issues list to piece together the concepts. It would be great if someone from the webpack team would review my config and comment on my use of webpack 2. I promise to maintain this example in accordance with your feedback.

To illustrate the importance of good examples, let me describe the latest issue that I am running into. It's the use of `resolve.modules`. In `webpack-common.config.js`, I have:

    resolve: {
        modules: [SRC_PATH, 'node_modules'],
        ...
    },

This allows me to specify all my imports relative to the `src` folder, instead of paths like `../../../xyz.js`. This is working great of JavaScript files, but I am running into a snag with css files. While `url()` and `composes` are picking up the correct modules starting from the `src` folder:

    background: url(assets/images/github-icon.png);
    composes: flexRowContainer from 'assets/styles/layout.css';

`@import` is not working as expected. I have to specify the full relative path:

    @import '../../../assets/styles/colors.css';

I don't even know if this is an issue with webpack, or css-loader or something else.

In summary, I think it is important to maintain some good quality examples as webpack 2 development proceeds. In absence of docs, this would be a great resource for people to continue working with webpack 2, minimizing the time wasted on trial and error.

**Comment by Kay2dan** [comment]
I seem to be having the same issue of modules...

https://github.com/webpack/webpack/issues/3397

**Comment by btipling** [comment]
All of these examples still use `loaders` as a top level under modules which is what webpack 2 has deprecated. That's just the webpack 1 way of doing things which still works. Webpack 2 uses `rules`, that replaces `loaders`. I hope nobody is using these examples just because they happened to work. They work because of temporary backwards compatibility, not because that's how webpack 2 works.

Examples of webpack 2 style:

http://javascriptplayground.com/blog/2016/10/moving-to-webpack-2/

https://github.com/vuejs/vue-loader/blob/master/docs/en/configurations/extract-css.md#webpack-2x-210-beta25

**Comment by marquizzo** [comment]
The webpack team just released new documentation for Webpack 2, which has some [good examples of the config file](https://webpack.js.org/concepts/configuration/)

**Comment by nitrog7** [comment]
Just updated our skeleton for use with Webpack 2. This skeleton should have plenty of examples for you to use: https://github.com/nitrogenlabs/arkhamjs-skeleton

**Comment by bultas** [comment]
I created template for simple JS apps where I using Webpack 2, Webpack-dev-server and Babel. 
Maybe it could be useful to someone who is looking for foundations..

https://github.com/bultas/minimal

**Comment by gg87** [comment]
Thank you @bultas !
You saved my whole day

**Comment by prateeksharmamay** [comment]
I found this tutorial to be really helpful: 
https://www.youtube.com/watch?v=GU-2T7k9NfI&list=PL55RiY5tL51rcCnrOrZixuOsZhAHHy6os
